src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,482 @@
+            if (fields[i].mutable_ == Immutable) {
+              noteSubtype(fields[i].type, superFields[i].type);
+            }
+          }
+        } else if (type.isArray()) {
+          auto elem = type.getArray().element;
+          if (elem.mutable_ == Immutable) {
+            noteSubtype(elem.type, super.getArray().element.type);
+          }
+        } else {
+          assert(type.isSignature());
+          auto sig = type.getSignature();
+          auto superSig = super.getSignature();
+          noteSubtype(superSig.params, sig.params);
+          noteSubtype(sig.results, superSig.results);
+        }
+      }
+
+      // TODO: This is expensive. Analyze casts incrementally after we
+      // initially analyze them.
+      analyzeCasts();
+    }
+  }
+
+  void analyzeCasts() {
+    // For each cast (src, dest) pair, any type that remains a subtype of src
+    // (meaning its values can inhabit locations typed src) and that was
+    // originally a subtype of dest (meaning its values would have passed the
+    // cast) should remain a subtype of dest so that its values continue to pass
+    // the cast.
+    //
+    // For every type, walk up its new supertype chain to find cast sources and
+    // compare against their associated cast destinations.
+    for (auto& [type, _] : typeDepths) {
+      for (auto superIt = supertypes.find(type); superIt != supertypes.end();
+           superIt = supertypes.find(superIt->second)) {
+        auto super = superIt->second;
+        auto destsIt = castTypes.find(super);
+        if (destsIt == castTypes.end()) {
+          continue;
+        }
+        for (auto dest : destsIt->second) {
+          if (HeapType::isSubType(type, dest)) {
+            noteSubtype(type, dest);
+          }
+        }
+      }
+    }
+  }
+
+  void optimizeTypes(Module& wasm) {
+    struct Rewriter : GlobalTypeRewriter {
+      Unsubtyping& parent;
+      Rewriter(Unsubtyping& parent, Module& wasm)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+      std::optional<HeapType> getSuperType(HeapType type) override {
+        if (auto it = parent.supertypes.find(type);
+            it != parent.supertypes.end() && !it->second.isBasic()) {
+          return it->second;
+        }
+        return std::nullopt;
+      }
+    };
+    Rewriter(*this, wasm).update();
+  }
+
+  void visitFunction(Function* func) {
+    if (func->body) {
+      noteSubtype(func->body->type, func->getResults());
+    }
+  }
+  void visitGlobal(Global* global) {
+    if (global->init) {
+      noteSubtype(global->init->type, global->type);
+    }
+  }
+  void visitElementSegment(ElementSegment* seg) {
+    if (seg->offset) {
+      noteSubtype(seg->type, getModule()->getTable(seg->table)->type);
+    }
+    for (auto init : seg->data) {
+      noteSubtype(init->type, seg->type);
+    }
+  }
+  void visitNop(Nop* curr) {}
+  void visitBlock(Block* curr) {
+    if (curr->name) {
+      BranchUtils::BranchSeeker seeker(curr->name);
```

Comment:
This will be quadratic. Maintaining a stack of blocks would be more efficient. Using ControlFlowWalker might be a simple way to get such a stack.